### PR TITLE
fix: node skips some epoch length txs making bridgeState corrupt

### LINKED ISSUE #340: Cannot increase epoch length
# Bounty

Probably related to #310.

To reproduce consider the following steps:
- Run the local env (integration tests with yarn start). This will spin up two validator network with epoch length of 2.
- Try to change epoch length to 3: submit and finalize a proposal to MinGov.
- You get leap-node:error Wrong epoch length in logs and epoch length stays 2

## Scope

- try to reproduce the case with unit test
- create an integration test to reproduce the issue
- fix the code (probably checkEpochLength handler)

## Deliverables

- updated code
- (possibly) updated integration tests
- (possibly) updated tests

## Reviewer checklist
- [ ] review the code for logical correctness, inefficiencies and style (TBD)
- [ ] ensure unit tests are passing
- [ ] ensure code coverage is not reduced
- [ ] ensure integration tests are passing for the PR branch

## Gain for the project

Working multi-validator setup

## Roles
bounty gardener: @troggy / 50 DAI
bounty worker: @troggy / 80%
bounty reviewer: name / 20%

## Funded by 

Dev Circle

### PR DESCRIPTION
**Pull request by troggy**
Requires: leapdao/leap-core#155, #348
Resolves #340 
Test to reproduce with: leapdao/integration-tests#69

## Tendermint replay protection

Tendermint has a [build-in replay protection](https://tendermint.com/docs/app-dev/app-development.html#replay-protection) which will reject the same tx from being added to the mempool twice. It seems to work by comparing raw tx bytes. Should two transactions have the same raw bytes, the second one will not make it to the mempool.

Now, EpochLength transaction in leap-core is merely a `[type,epochLength]` tuple. `Tx.epochLength(X)` will have the same raw bytes for the same `X`. Consider the following pretty valid transaction flow:

```
1. Tx.epochLength(2)
2. Tx.epochLength(4)
3. Tx.epochLength(2)
```

The third transaction will be rejected by tendermint's replay protection.

## How it affects leap-node

The node listens for root chain events, including `EpochLength` event. Everytime event comes, the node
a) keeps record of the epoch length from event (in the `bridgeState.epochLengths` list). [bridgeState](https://github.com/leapdao/leap-node/blob/5195438c54b7c2670e7314d102b690255b4a203b/src/bridgeState.js#L135)
b) creates `epochLength` transaction and relays it. [eventsRelay](https://github.com/leapdao/leap-node/blob/5195438c54b7c2670e7314d102b690255b4a203b/src/eventsRelay.js#L80)
c) handles `epochLength` tx, once received through ABCI from tendermint. [checkEpochLength](https://github.com/leapdao/leap-node/blob/5195438c54b7c2670e7314d102b690255b4a203b/src/tx/applyTx/checkEpochLength.js#L10)

The step (c) expects all the a,b,c steps always executed for every `EpochLength` event. However, if the node receives two `EpochLength(X)` events for the same epoch length, only steps (a) and (b) will be executed. Step (c) would never happen for the second event, because the transaction created on (b) will be rejected by Tendermint's replay protection. In the end, we have `bridgeState.epochLength` mutated by (a), which will make `checkEpochLength` fail for any further epoch length tx.

## Proposed solution

Add more differentiation to EpochLength transactions. I suggest adding blockHeight to it. See: https://github.com/leapdao/leap-core/pull/155

These two things needs to be merged first for the fix to work: leapdao/leap-core#155, #348

### CODE REVIEW

**Comment by troggy** [comment]
temporary using leap-core@0.36.2-beta which includes the fix https://github.com/leapdao/leap-core/pull/155

**Comment by sunify** [comment]
Holy shit. I wonder how we ain't dead yet :D

I have a feeling we should create the new tx type and rename the old one to «Tx.epochLength_legacy» or something like that. If this PR will merged it'll make sync with testnet or mainnet impossible — Tx.fromRaw will fail with `malformed epoch tx.` error

**Comment by troggy** [comment]
> Tx.fromRaw will fail with malformed epoch tx. error

great catch, totally overlooked this.

File: src/eventsRelay.js [hunk 1]
```diff
@@ -75,9 +75,9 @@ module.exports = class EventsRelay {
         );
         this.sendDelayed(tx);
       },
-      EpochLength: async event => {
+      EpochLength: async (event) => {
         const { epochLength } = event.returnValues;
-        const tx = Tx.epochLength(Number(epochLength));
+        const tx = Tx.epochLength(Number(epochLength), Number(event.blockNumber));
```
Comment (sunify) [inline src/eventsRelay.js]: We can save blockNumber in `bridgeState.epochLengths` as well and use it for tx validation 🤔

Comment (troggy) [reply src/eventsRelay.js]: I didn't get the idea. You propose to use these blockNumbers here when creating epochLength tx ?

**Comment by troggy** [comment]
> We have almost identical (in terms of code) tx type MIN_GAS_PRICE and It affected by the same problem. We can create the new issue or fix here — up to you

👍
Opened as issue for that: https://github.com/leapdao/leap-core/issues/156

File: src/eventsRelay.js (hunk 1 above)
Comment (sunify) [reply src/eventsRelay.js]: In bridgeState we [store all epochLength changes](https://github.com/leapdao/leap-node/blob/master/src/bridgeState.js#L134-L136). We can add blockNumber into this array (`[epochLength, blockNumber][]`) and use it in [this check](https://github.com/leapdao/leap-node/blob/fix/340-epoch-length/src/tx/applyTx/checkEpochLength.js)

Btw, do not forget to update check anyway — it should handle two tx types (now it handles zero :-))

**Comment by troggy** [comment]
> If this PR will merged it'll make sync with testnet or mainnet impossible — Tx.fromRaw will fail with malformed epoch tx. error

addressed with new tx type. I've quickly checked with testnet sync — the sync goes beyond the height 4, no error.

File: src/eventsRelay.js (hunk 1 above)
Comment (troggy) [reply src/eventsRelay.js]: hm.. I've updated the check.

I don't see why we need to store blockNumber in bridgeState — we can just take it from the transaction (like I do in my code). If there is no blockNumber in epochLength, chance are that `checkEpochLength` won't be called at all, exactly the issue we are fixing here.

Comment (troggy) [reply src/eventsRelay.js]: or you mean, we can compare height from bridgeState with the one from tx ?

Comment (sunify) [reply src/eventsRelay.js]: Yes :-)

**Pull request merged**